### PR TITLE
Add new PHP format

### PR DESCRIPTION
**Pull request by swissspidy**
## What?

This adds a new `.php` _output_ format for translations.

> [!NOTE]  
>  Import is not yet implemented. It requires something like PHP-Parser to do that safely, as you don't want to include/require a random user-provided PHP file. This is best done in a separate PR.

## Why?

The WordPress core performance team is working on a [Performant Translations feature plugin](https://wordpress.org/plugins/performant-translations/) that uses a new approach to handle translation files in WordPress, making localization blazing fast. This plugin helps to make localized WordPress sites faster by replacing the traditional MO translation files with PHP files, which are much faster to parse. 

Learn more at https://make.wordpress.org/core/2023/09/05/call-for-testing-performant-translations/

Ideally, these PHP files are served directly from translate.wordpress.org, which is why support needs to be added to GlotPress first.

See https://meta.trac.wordpress.org/ticket/7296 for enabling that on the dotorg side of things.

## How?

Translations are exported using an optimized version of `var_export()` to use short array notation. A polyfill for `array_is_list` is included. Both are used to keep the file size as small as possible.

See https://make.wordpress.org/core/2023/09/05/call-for-testing-performant-translations/ and https://github.com/swissspidy/performant-translations

## Testing Instructions

1. Export translations as PHP
2. Check the syntax to see it's valid
3. Include/require that file in PHP, ideally in WP using that feature plugin.

## Screenshots or screencast

### CODE REVIEW

**Review by amieiro — COMMENTED** [review]
Before reviewing this PR, I reviewed the existing format headers.

- Android. XML file (in a XML comment):
``` 
<!--
Translation-Revision-Date: 2023-09-26 13:29:05+0000
Plural-Forms: nplurals=2; plural=n != 1;
Generator: GlotPress/4.0.0-alpha.11
Language: gl_ES
-->
```
- Jed 1.x. JSON file:
```
{
  "translation-revision-date": "2023-09-26 13:29:05+0000",
  "generator": "GlotPress/4.0.0-alpha.11",
  "domain": "messages",
  "locale_data": {
    "messages": {
      "": {
        "domain": "messages",
        "plural-forms": "nplurals=2; plural=n != 1;",
        "lang": "gl_ES"
      },
```
- JSON format. JSON file. Doesn't have any header.
- NGX-Translate. JSON file. Doesn't have any header.
- Portable Object Message Catalog (.po/.pot)
 ```
 "PO-Revision-Date: 2023-09-26 13:29:05+0000\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=n != 1;\n"
"X-Generator: GlotPress/4.0.0-alpha.11\n"
"Language: gl_ES\n"
"Project-Id-Version: Plugins - Elementor\n"
 ```
- Java Properties File (.properties). Comments, starting the line with #.
```
# Translation-Revision-Date: 2023-09-26 13:29:05+0000
# Plural-Forms: nplurals=2; plural=n != 1;
# Generator: GlotPress/4.0.0-alpha.11
# Language: gl_ES
```
- .NET Resource (.resx). 

```
	<resheader name="resmimetype">
		<value>text/microsoft-resx</value>
	</resheader>
	<resheader name="version">
		<value>2.0</value>
	</resheader>
	<resheader name="reader">
		<value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
	</resheader>
	<resheader name="writer">
		<value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
	</resheader>
	<resheader name="translation_revision_date">
		<value>2023-09-26 13:29:05+0000</value>
	</resheader>
	<resheader name="plural_forms">
		<value>nplurals=2; plural=n != 1;</value>
	</resheader>
	<resheader name="generator">
		<value>GlotPress/4.0.0-alpha.11</value>
	</resheader>
	<resheader name="language">
		<value>gl_ES</value>
	</resheader>
```
- Mac OS X / iOS Strings File (.strings)

```
/* Translation-Revision-Date: 2023-09-26 13:29:05+0000 */
/* Plural-Forms: nplurals=2; plural=n != 1; */
/* Generator: GlotPress/4.0.0-alpha.11 */
/* Language: gl_ES */
```

### New format. Headers

```
'x-generator' => 'GlotPress/4.0.0-alpha.11', 
'translation-revision-date' => '2023-09-29 17:24:57+0000', 
'plural-forms' => 'nplurals=2; plural=n != 1;',
```

I will propose these small improvements in the header:
- Add the `language` item. E.g. `'language' => 'gl_ES'`.
- Add `project-id-version` item to be able to easily identify the project related with this translation file. E.g. `'project-id-version' => Plugins - Elementor`.

### Importing new files (.php extension)

If a user tries to import a .php file with this new format, she gets an error with the text "Couldn't load translations from file!", because this PR didn't have this code implemented.

image

I suggest an improvement, filtering this type of import file in the dropdown element, so the used won't see any PHP format to import.

In the `gp-templates/project-import.php` I suggest this improvement:

Replace this code:

```
	foreach ( GP::$formats as $slug => $format ) {
```

with 

```
	$formats = GP::$formats;
	unset( $formats['php'] ); // Remove PHP format from the list of formats until we have an import process.
	foreach ( $formats as $slug => $format ) {
```

image

### Testing the functionality

To test this functionality:
- In a local GlotPress, I created a tree structure with a `Plugins` project where I store all plugins: Projects -> Plugins -> Elementor.
- I downloaded the .po file with the translations from the w.org site for [Elementor in Galician language](https://translate.wordpress.org/projects/wp-plugins/elementor/stable/gl/default/) and imported this file.
- I created the .php file with the translations and change the name to the same used by the [Performant Translations](https://wordpress.org/plugins/performant-translations/) plugin: `elementor-gl_ES.mo.php`.


In a new WordPress installation:
- I remove all the plugins. 
- I installed the [Performant Translations](https://wordpress.org/plugins/performant-translations/) plugin, to be able to use it with WordPress, as a regular translation format.
- I installed the [Query Monitor](https://wordpress.org/plugins/query-monitor/) plugin, to be able to see the loaded translations.
- I installed the [Elementor](https://wordpress.org/plugins/elementor/), to review its translations.
- I overwrite the .php file (elementor-gl_ES.mo.php) in the `wp-content/languages/plugins` folder from GlotPress.
- I checked the translations in the Elementor backend and in the Query Monitor (to see the file used) and it works fine.

**Comment by swissspidy** [comment]
> I suggest an improvement, filtering this type of import file in the dropdown element, so the used won't see any PHP format to import.

Good point! I also updated `gp_get_import_file_format()` accordingly so that you'll get a better error message when choosing "Auto" for the import.

> I will propose these small improvements in the header:
> 
> * Add the `language` item. E.g. `'language' => 'gl_ES'`.
> * Add `project-id-version` item to be able to easily identify the project related with this translation file. E.g. `'project-id-version' => Plugins - Elementor`.

Done 👍

File: gp-includes/formats/format-php.php [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ */
+class GP_Format_PHP extends GP_Format {
+	/**
+	 * Name of file format, used in file format dropdowns.
+	 *
+	 * @since 4.0.0
+	 *
+	 * @var string
+	 */
+	public $name = 'PHP (.php)';
```
Comment (amieiro) [inline gp-includes/formats/format-php.php]: As the [Performant Translations](https://wordpress.org/plugins/performant-translations/) plugin uses the ".mo.php" extension, what do you think about changing:
```
public $name = 'PHP (.php)';
public $extension = 'php';
```
to
```
public $name = 'PHP (.mo.php)';
public $extension = 'mo.php';
```
because we are going to add the "mo." to each file when we automatically generate them.

File: gp-includes/formats/format-php.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+			$current        = GP::$project->get( $current->parent_project_id );
+			$project_tree[] = $current->name;
+		}
+
+		$project_tree = array_reverse( $project_tree );
+
+		$project_id_version = implode( ' - ', $project_tree );
+
+		/** This filter is documented in gp-includes/formats/format-pomo.php */
+		$project_id_version = apply_filters( 'gp_pomo_export_project_id_version', $project_id_version, $project_tree );
```
Comment (amieiro) [inline gp-includes/formats/format-php.php]: Related with this filter, it was thought to be used in the `pomo`, not for the `php` export.

What do you think about creating a new filter, only for the PHP export?

```suggestion
		/**
		 * Filter the project name and version header before export.
		 *
		 * @since 4.0.0
		 *
		 * @param string $project_id_version The default project name/version to use in the header and
		 *                                   comments ( "Parent - Child - GrandChild - etc." by default).
		 * @param array  $project_tree       An array of the parent/child project tree, ordered from Parent
		 *                                   to child to grandchild to etc...
		 */
		$project_id_version = apply_filters( 'gp_php_export_project_id_version', $project_id_version, $project_tree );
```

File: gp-includes/formats/format-php.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			'translation-revision-date' => GP::$translation->last_modified( $translation_set ) . '+0000',
+			'plural-forms'              => "nplurals=$locale->nplurals; plural=$locale->plural_expression;",
+			'project-id-version'        => $project_id_version,
+			'language'                  => $language_code,
+			'messages'                  => array(),
+		);
+
+		/* @var Translation_Entry $entry */
+		foreach ( $entries as $entry ) {
+			$key = $entry->context ? $entry->context . chr( 4 ) . $entry->singular : $entry->singular;
```
Comment (amieiro) [inline gp-includes/formats/format-php.php:91]: In the 			
```
$key = $entry->context ? $entry->context . chr( 4 ) . $entry->singular : $entry->singular;
``` 
you are adding the context for the original string. 

Here we can take 2 approaches:

1. Avoid all extra information, using this new format as only an export format, removing the `context`.
2. Add all the elements in the `Translation_Entry`  class:

```
*     @type string $context             A string differentiating two equal strings
*                                       used in different contexts.
*     @type string $translator_comments Comments left by translators.
*     @type string $extracted_comments  Comments left by developers.
*     @type array  $references          Places in the code this string is used, in
*                                       relative_to_root_path/file.php:linenum form.
*     @type array  $flags               Flags like php-format.
```
I think this second approach is the good one, because it will have all the information to use this new format as import format in the future.

File: gp-includes/formats/format-php.php (hunk 1 above)
Comment (swissspidy) [reply gp-includes/formats/format-php.php]: Sounds good 👍 

Note that the file extension is not 100% defined yet, there was also a suggestion to use `.l10n.php` or similar.

I'll update it once it's settled.

File: gp-includes/formats/format-php.php (hunk 3 above)
Comment (swissspidy) [reply gp-includes/formats/format-php.php:91]: Hmm not sure about this one.

I see the PHP format more as a lightweight export-only format that doesn't necessarily have to retain all of the metadata.

(1) is not an option. Context is a requirement for being able to translate strings with context, as it makes things unique. You can't omit this. Otherwise how would you store translations for `_x( 'Run', 'noun' )` and `_x( 'Run', 'verb' )`? So this is required.

References and comments on the other hand are optional metadata for translators. Including those in the export by default would significantly increase file size. It would also mean not being able to use `implode()` for the messages, as then it would need to be an array (e.g. `$result['messages'][ $key ] = [ 'translation' => ..., 'comments' => ..., 'references' => ... ]`, which would cause memory overhead as well due to all the array allocations.

In short, this would negate many of the performance benefits of the current, extremely lightweight format.

File: gp-includes/formats/format-php.php (hunk 1 above)
Comment (amieiro) [reply gp-includes/formats/format-php.php]: Ok. If it is ok to you, we can merge this PR with the current extension and when you will have the final extension, we can update it in a new PR.

File: gp-includes/formats/format-php.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+		while ( $current->parent_project_id > 0 ) {
+			$current        = GP::$project->get( $current->parent_project_id );
+			$project_tree[] = $current->name;
+		}
+
+		$project_tree = array_reverse( $project_tree );
+
+		$project_id_version = implode( ' - ', $project_tree );
+
+		/** This filter is documented in gp-includes/formats/format-pomo.php */
```
Comment (amieiro) [inline gp-includes/formats/format-php.php]: Can we remove this comment, as the new filter is documented in the current file?
```suggestion
```

File: gp-includes/formats/format-php.php (hunk 3 above)
Comment (amieiro) [reply gp-includes/formats/format-php.php:91]: Ok, this explanation makes sense to me, so we will use this new .php format as export-only format.

**Review by amieiro — APPROVED** [review]
